threads: avoid deadlock from recursive lock acquire

vtjnash opened the pull request:
Finalizers can't safely acquire many essential locks (such as the
iolock, to cleanup libuv objects) if they are run inside another lock.
Therefore, inhibit all finalizers on the thread until all locks are
released (previously, this was only true for our internal locks).

vtjnash committed:
threads: avoid deadlock from recursive lock acquire

Finalizers can't safely acquire many essential locks (such as the
iolock, to cleanup libuv objects) if they are run inside another lock.
Therefore, inhibit all finalizers on the thread until all locks are
released (previously, this was only true for our internal locks).

JeffBezanson commented on base/locks-mt.jl:
```diff
@@ -61,10 +61,12 @@ Base.assert_havelock(l::SpinLock) = islocked(l) ? nothing : Base.concurrency_vio
 function lock(l::SpinLock)
     while true
         if _get(l) == 0
+            GC.enable_finalizers(false)
```
Why can't this be done just before returning?

vtjnash replied:
The entire point of this bug fix is that it must be done before lock acquisition

JeffBezanson replied:
For comparison, `jl_mutex_lock` only disables finalizers after `jl_mutex_wait` returns. I thought that would correspond to disabling them just after `if p == 0` here.

vtjnash replied:
jl_mutex_lock can get away with doing potentially suspect things because it can ensure they are indistinguishable to the user

maleadt reviewed (approved):
Works fine with CUDA.jl (which had its own version of this).

vchuravy commented:
The failure on Linux32 bit is related:

```
Error in testset compiler/codegen:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/compiler/codegen.jl:251
  Expression: was_gced
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:130
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:131
  Expression: lock(enable_finalizers_count, l) == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:132
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:136
  Expression: enable_finalizers_count() == 2
   Evaluated: 3 == 2
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:138
  Expression: enable_finalizers_count() == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:140
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:142
  Expression: enable_finalizers_count() == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:144
  Expression: enable_finalizers_count() == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:146
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:667
  Expression: contains_warn(read(fname, String), $(Expr(:escape, "WARNING: GC finalizers already enabled on this thread.")))
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:130
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:131
  Expression: lock(enable_finalizers_count, l) == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:132
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:136
  Expression: enable_finalizers_count() == 2
   Evaluated: 3 == 2
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:138
  Expression: enable_finalizers_count() == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:140
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:142
  Expression: enable_finalizers_count() == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:144
  Expression: enable_finalizers_count() == 1
   Evaluated: 2 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:146
  Expression: enable_finalizers_count() == 0
   Evaluated: 1 == 0
Error in testset misc:
Test Failed at /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:667
  Expression: contains_warn(read(fname, String), $(Expr(:escape, "WARNING: GC finalizers already enabled on this thread.")))
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:165
  Expression: c[] == 1
   Evaluated: 0 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:167
  Expression: c[] == 1
   Evaluated: 0 == 1
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:169
  Expression: c[] == 100
   Evaluated: 99 == 100
Error in testset misc:
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/misc.jl:505
  Expression: finalized
```

vtjnash commented:
Okay, I changed the implementation, as adding an error check for that revealed a number of violators. Should be happy now?

vtjnash committed:
move timing_stack field from jl_task_t onto the stack also

vtjnash commented:
Okay, I changed the implementation, as that was not sufficient. Should be happy now?

Unrelated, I also reduced the size of jl_task_t here by 2 pointers, to be consistent in the implementation.

vchuravy commented:
> Should be happy now?

Linux 64bit is still not happy.

vtjnash commented:
Okay, I changed the implementation, as I missed a line of code, and added some assertions for it. Should be happy now?

Will squash now, as CI is green.

JeffBezanson commented on src/gc.c:
```diff
@@ -392,12 +392,36 @@ static void run_finalizers(jl_ptls_t ptls)
     arraylist_free(&copied_list);
 }
 
+JL_DLLEXPORT int jl_gc_get_enable_finalizers(jl_ptls_t ptls)
```
```suggestion
JL_DLLEXPORT int jl_gc_get_finalizers_inhibited(jl_ptls_t ptls)
```

vtjnash replied:
I named it after the method name, rather than the field. But it's exposed as a test-only API currently, partly because it seemed like it shouldn't have a real use and need exporting, but also because I didn't want to name it

JeffBezanson replied:
Yes, I see it's test-only, but it's still confusing for it to return the opposite of what the name seems to say.
